document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresDatastore.java

```diff
@@ -82,6 +95,8 @@ public Set<String> listCollections() {
 
   @Override
   public boolean createCollection(String collectionName, Map<String, String> options) {
+    final PostgresTableIdentifier identifier = PostgresTableIdentifier.parse(collectionName);
+    identifier.getSchema().ifPresent(this::createSchemaIfNotExists);
```

Comment:
- If the collectionName is 'x.y', are we considering 'x' as the schema and 'y' as the table name?"

- Would there be a case of the collection name being 'x.y.z'? In that case, the first part would be the schema, and 'y.z' would be the tableName?

So far, we were considering all the tables under `public` schema.

Reply:
> If the collectionName is 'x.y,' are we considering 'x' as the schema and 'y' as the table name? I see this in a PR comment too."

Yes

>Would there be a case of the collection name being 'x.y.z'? 

Could be. Mongo puts no restrictions. In this change this is now supported.

> In that case, the first part would be the schema, and 'y.z' would be the tableName. So, if we could support a dot as part of the table name, I think we should put them in the default schema public as we used to do before. This impl based logical grouping can lead to a higher number of random schemas and can caused backup/restore policy difficult in case of shared instance.

Initially went with this approach. The reason I didn't is that there are certain queries that do need to cross schemas. One of the bugs that motivated this change was the lack of postgres metrics, fixed in https://github.com/hypertrace/document-store/pull/181/commits/4dd0334965f700b1df6af4c6ef5eca67e8577a1e 

We could have introduced some postgres-specific way of specifying a schema separately from the collection I suppose, but I assumed our unit of backup/restore was the database rather than the schema - is that not the case?

> By the way, I had a thought of moving all collections under the 'hypertrace' schema, so it makes it easier for backup or any policies to apply if its used with shared instance with shared db. Well, for now, we are creating new db, so we can fully restore that db.

I guess you partially answer my last question here. What advantages would there be to using a schema over the current database approach? IMO, I like different services being able to maintain distinct schemas that are impl details.

Reply:
> > If the collectionName is 'x.y,' are we considering 'x' as the schema and 'y' as the table name? I see this in a PR comment too."
> 
> Yes
> 
> > Would there be a case of the collection name being 'x.y.z'?
> 
> Could be. Mongo puts no restrictions. In this change this is now supported.
> 
> > In that case, the first part would be the schema, and 'y.z' would be the tableName. So, if we could support a dot as part of the table name, I think we should put them in the default schema public as we used to do before. This impl based logical grouping can lead to a higher number of random schemas and can caused backup/restore policy difficult in case of shared instance.
> 
> Initially went with this approach. The reason I ended up using the schema instead is that there are certain queries that do need to cross schemas. One of the bugs that motivated this change was the lack of postgres metrics, fixed in [4dd0334](https://github.com/hypertrace/document-store/commit/4dd0334965f700b1df6af4c6ef5eca67e8577a1e)
> 
> We could have introduced some postgres-specific way of specifying a schema separately from the collection I suppose, but I assumed our unit of backup/restore was the database rather than the schema - is that not the case?
> 
> > By the way, I had a thought of moving all collections under the 'hypertrace' schema, so it makes it easier for backup or any policies to apply if its used with shared instance with shared db. Well, for now, we are creating new db, so we can fully restore that db.
> 
> I guess you partially answer my last question here. What advantages would there be to using a schema over the current database approach? IMO, I like different services being able to maintain distinct schemas that are impl details.

Yes, currently, we are conducting maintenance at the database level. Agree that this approach allows us to logically group tables under a given database.

Since tables are created dynamically through the collection API, the corresponding schema is also created dynamically. Without a corresponding `deleteSchema API`, this could lead to dangling schemas and could introduce manual overhead.

As of now, we do not delete any collections except in test cases, so we can proceed. Later, we can extend the implementation of the `deleteCollection API` on the PostgreSQL side. If there are no tables left under the given non-public schema, we can then drop that schema as well.

Reply:
Makes sense. I considered the delete case too and skipped with similar thoughts.